src/frontend/src/optimizer/rule/batch/batch_push_limit_to_scan_rule.rs

```diff
@@ -0,0 +1,54 @@
+//  Copyright 2023 RisingWave Labs
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+// This source code is licensed under both the GPLv2 (found in the
+// COPYING file in the root directory) and Apache 2.0 License
+// (found in the LICENSE.Apache file in the root directory).
+
+use itertools::Itertools;
+
+use crate::optimizer::plan_node::generic::PhysicalPlanRef;
+use crate::optimizer::plan_node::{BatchLimit, BatchSeqScan, PlanTreeNodeUnary};
+use crate::optimizer::rule::{BoxedRule, Rule};
+use crate::optimizer::PlanRef;
+
+pub struct BatchPushLimitToScanRule {}
+
+impl Rule for BatchPushLimitToScanRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let limit: &BatchLimit = plan.as_batch_limit()?;
+        let limit_input = limit.input();
+        let scan: &BatchSeqScan = limit_input.as_batch_seq_scan()?;
+        if scan.limit().is_some() {
+            return None;
+        }
+        let pushed_limit = limit.limit() + limit.offset();
+        let mut scan_core = scan.core().clone();
+        scan_core.chunk_size = Some((u32::MAX as u64).min(pushed_limit) as u32);
```

Comment:
What if `limit + offset` is very large?

Reply:
The `chunk_size` will be limited by configuration `chunk_size` on the executor side. BTW, I want to deprecate this method later.